compiler/codegen/expr.cpp

```diff
@@ -2408,31 +2285,36 @@ GenRet codegenArgForFormal(GenRet arg,
   return arg;
 }
 
-// if fSym is non-NULL, we use that to decide what to dereference.
+// if fn is non-NULL, we use that to decide what to dereference.
 // Otherwise, if defaultToValues=true, we will codegenValue() the arguments,
 //            and if it is false, they will pass by reference if they
 //            are references.
 static
 GenRet codegenCallExpr(GenRet function,
                        std::vector<GenRet> & args,
-                       FnSymbol* fSym,
+                       FnSymbol* fn,
+#ifdef HAVE_LLVM
+                       clang::FunctionDecl* FD,
+#else
+                       void* FD,
+#endif
```

Comment:
I wish you handled this differently. I feel that preprocessor macros in function signatures are a bit too much.

How about doing this same logic as a whole in the function signature, as you have done in the declaration of this function in the beginning of expr.cpp? I am aware that there'll be some code duplication but, that might make it cleaner.

Another alternative that I can think of is `typedef` ing `FunctionDecl`.


I think we are not passing non-NULL `fn` and `FD` at the same time, so, maybe ideally we can create two separate overloads of this, where one is within `#ifdef HAVE_LLVM`, but I am not asking you to refactor `codegenCallExpr`! :)

If whatever you end up doing doesn't look cleaner to you, then let's not worry about this for now.

Reply:
I'll look at adding a typedef, which will just be within this file.